### ISSUE TITLE
`import` best practices

### ISSUE THREAD

**Issue by nathanbuchar** [lexmag/elixir-style-guide#31]
I have a few questions about best practices when using `import`, and the like.

When using `import`, should I simply import the entire module and not bother with the `:only` option?

```elixir
# ...

import App.Web.SessionHelpers
import Comeonin.Bcrypt
```

This approach certainly looks the cleanest, but what if the module being imported has a large number of public functions? With that I risk polluting the namespace with a bunch of unused functions. It makes sense to only import what one needs, yes?

Additionally, when a function is called within a module which was imported from a different module, how will I or future maintainers know where it came from (easily, anyway)? I know I've seen this in code before, I'll see a function like `clear_user_password(user)` and the function doesn't exist as a function within the current module and there are like ten imports (oy!). Using `:only` can help point future maintainers in the right direction, as it makes what is used and from where very clear.

```elixir
# ...

import App.Web.SessionHelpers, only: [clear_current_user: 1, get_current_user_safe: 1, set_current_user: 2 ]
import Comeonin.Bcrypt, only: [checkpw: 2]
```

But like above, this can easily get out of hand. So is it better in this case to stack them, like below?

```elixir
# ...

import App.Web.SessionHelpers, only: [
  clear_current_user: 1,
  get_current_user_safe: 1,
  set_current_user: 2,
]

import Comeonin.Bcrypt, only: [
  checkpw: 2,
]
```
\
Or is there no "right" way, because it depends heavily on the context. E.g., if I'm using the majority of the functions from an imported module I should just import it without `:only`, but use `:only` if I'm just borrowing a function or two?

Please advise, thank you :)

**Comment by whatyouhide** [comment]
@nathanbuchar there is no right way, it depends on the context. To be fair this is an issue that is borderline style-related, as it's mostly about semantics of imports (while this style guide is mostly about visual style).

For completeness, my opinion is this: use `import` as sparingly as you can. When you use it, try to use `:only` as much as you can. If you have to import all functions in the current context (first example that comes to mind is `import Ecto.Query`), it's fine not to use `:only`, but things like `Ecto` are a very special case because they're very well known and very established libraries, so people know their API and the things that they expose. For something like `Comeonin.Bcrypt`, my suggestion is to use the fully qualified name `Comeonin.Bcrypt.checkpw(...)`. If some function of yours uses it many times, you can locally import with `import Comeonin.Bcrypt, only: [checkpw: 2]` limiting it to the scope of the function. If you use it many times in the same file, it might be fine to import it module-wide.

TL;DR: try not to import ever, if you do try to import smallest possible amount of functions, if you have to import all make sure it's a well known API (`Ecto.Query`, `Plug.Conn`).

Closing this issue anyways since there are no really style-related questions. Thanks @nathanbuchar! 💟

**Comment by nathanbuchar** [comment]
Great, thanks for the tips :)
